Quotes can crash due to missing implicit type instance in local function

Regression reproduced based on the failure in [Open CB #4763](https://scala3.westeurope.cloudapp.azure.com/job/buildCommunityProject/4763/). Problem found in [playframework/anorm](https://github.com/playframework/anorm)

## Compiler version
Fails in 3.2.1-RC1
Fails in 3.2.0
Works in 3.1.3

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
package anorm.macros
sealed trait Row
sealed trait SqlResult[A]

import scala.quoted.{ Expr, Quotes, Type }

private[anorm] object RowParserImpl {
  def apply[A](using q:Quotes)(using Type[A]): Expr[Row => SqlResult[A]] = {
    import q.reflect.*

    inline def fails: Expr[SqlResult[A]] =
      Match(???, ???).asExprOf[SqlResult[A]] 
    
    // Workarounds:
    inline def compiles(using Type[SqlResult[A]]): Expr[SqlResult[A]] =
      Match(???, ???).asExprOf[SqlResult[A]] 

    // Compiles
    Match(???, ???).asExprOf[SqlResult[A]] 
    ???
  }
}
```

## Output (click arrow to expand)
<details>

```scala
exception occurred while compiling test.scala
java.lang.AssertionError: assertion failed while compiling test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
        at dotty.tools.dotc.ast.TypedTreeInfo.constToLiteral(TreeInfo.scala:565)
        at dotty.tools.dotc.ast.TypedTreeInfo.constToLiteral$(TreeInfo.scala:387)
        at dotty.tools.dotc.ast.tpd$.constToLiteral(tpd.scala:23)
        at dotty.tools.dotc.transform.FirstTransform.transformBlock(FirstTransform.scala:174)
        at dotty.tools.dotc.transform.MegaPhase.goBlock(MegaPhase.scala:740)
        at dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:450)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:298)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:459)
        at dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:459)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:303)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:278)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:459)
        at dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:459)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:279)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:320)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:248)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:425)
        at dotty.tools.dotc.transform.MegaPhase.loop$2(MegaPhase.scala:444)
        at dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:449)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:298)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:248)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:425)
        at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:359)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:255)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:425)
        at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:379)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:382)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:454)
        at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:466)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:311)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:312)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:234)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:245)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:253)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:262)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:262)
        at dotty.tools.dotc.Run.compileSources(Run.scala:186)
        at dotty.tools.dotc.Run.compile(Run.scala:170)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
        at dotty.tools.dotc.Driver.process(Driver.scala:195)
        at dotty.tools.dotc.Driver.process(Driver.scala:163)
        at dotty.tools.dotc.Driver.process(Driver.scala:175)
        at dotty.tools.dotc.Driver.main(Driver.scala:205)
        at dotty.tools.MainGenericCompiler$.run$1(MainGenericCompiler.scala:162)
        at dotty.tools.MainGenericCompiler$.main(MainGenericCompiler.scala:186)
        at dotty.tools.MainGenericCompiler.main(MainGenericCompiler.scala)
        ```
</details>

--- a/compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala
+++ b/compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala
@@ -113,12 +113,10 @@ class PickleQuotes extends MacroTransform {
             case _ =>
               val (contents, tptWithHoles) = makeHoles(tpt)
               PickleQuotes(quotes, tptWithHoles, contents, tpt.tpe, true)
-        case tree: DefDef if tree.symbol.is(Macro) =>
+        case tree: DefDef if tree.symbol.is(Macro) || tree.symbol.isInlineMethod =>
           // Shrink size of the tree. The methods have already been inlined.
           // TODO move to FirstTransform to trigger even without quotes
           cpy.DefDef(tree)(rhs = defaultValue(tree.rhs.tpe))
-        case _: DefDef if tree.symbol.isInlineMethod =>
-          tree
         case _ =>
           super.transform(tree)
   }

--- a/tests/pos-macros/i15985.scala
+++ b/tests/pos-macros/i15985.scala
@@ -1,0 +1,28 @@
+package anorm.macros
+sealed trait Row
+sealed trait SqlResult[A]
+
+import scala.quoted.{ Expr, Quotes, Type }
+
+private[anorm] object RowParserImpl {
+  def apply[A](using q:Quotes)(using a: Type[A]): Expr[Row => SqlResult[A]] = {
+    import q.reflect.*
+
+    inline def f1: Expr[SqlResult[A]] =
+      Match(???, ???).asExprOf[SqlResult[A]] // (using Type.of[anorm.macros.SqlResult[A]] })
+
+    inline def f2: Expr[SqlResult[A]] =
+      Match(???, ???).asExprOf[SqlResult[A]](using Type.of[SqlResult[A]])
+      // In Staging phase it becomes
+      //  ..asExprOf[..](using Type.of[{ @SplicedType type a$_$3 = a.Underlying; anorm.macros.SqlResult[a$_$3] }])
+
+    inline def f3(using Type[SqlResult[A]]): Expr[SqlResult[A]] =
+      Match(???, ???).asExprOf[SqlResult[A]]
+
+    f1
+    f2
+    f3
+
+    ???
+  }
+}

